### PR TITLE
Motion Matching: Added standard scaler

### PR DESCRIPTION
**Pull request by amzn-jillich**
The standard scaler can be used to normalize the feature matrix, the query vector and other data. It standardizes features by removing the mean and scaling to the unit variance. This implementation is mimicing the behavior of the standard scaler from scikit-learn (sklearn.preprocessing.StandardScaler). As we use floats by default, our errors are a bigger, especially if the variance is small as this leads to a division by a small value. In case the calculated standard deviation for a given feature is smaller than the given s_epsilon value, the standard deviation gets force-set to 1.0 to avoid divisions by infinity and preserve the value when doing the transform -> inverse-transform roundtrip.

If the standard deviation is small and basically a feature hardly changes (but does change a little bit more than our s_epsilon), the scaling is reaching floating point accuracy limits pretty quickly and the results differ from sklearn's StandardScaller as that internally seems to use double precision.

Signed-off-by: Benjamin Jillich <jillich@amazon.com>

### CODE REVIEW

File: Gems/MotionMatching/Code/Source/FeatureMatrixStandardScaler.h [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        }
+
+        static constexpr float s_epsilon = AZ::Constants::FloatEpsilon;
+
+        void SaveAsCsv(const char* filename, const AZStd::vector<AZStd::string>& columnNames = {});
+
+    private:
+        AZStd::vector<float> m_means; //!< The mean value for each feature / column.
+        AZStd::vector<float> m_standardDeviations; //!< The standard deviation for each feature / column.
+        static constexpr float s_deltaDegreesOfFreedom = 0.0f; // Used to set the bias of the estimator. [Default=unbiased]
```
Comment (greerdv) [inline Gems/MotionMatching/Code/Source/FeatureMatrixStandardScaler.h:62]: I think there is a bit of subtlety here depending on whether you're talking about the sample standard deviation or the estimate of the population standard deviation. When you talk about bias, it's about whether the expected error between your estimate of the population standard deviation and the ground truth is zero (unbiased) or non-zero (biased). If you set the DOF to 0, you actually will have a biased estimate of the population standard deviation (it will on average be slightly too small because on average the sample values will be slightly closer to the sample mean than they are to the population mean). To get an unbiased estimate of the population standard deviation, you actually would want DOF = 1. 

Having said all that:
(a) I think it's perfectly reasonable to scale based on the sample mean and standard deviation
(b) if the sample size is even remotely large, the difference is basically going to be negligible

To summarize, I think strictly speaking referring to DOF = 0 as unbiased is not correct. But since I think the effect is going to be insignificant, I wonder if you really need the DOF term at all or if it could just be completely omitted?

**Review by greerdv — APPROVED** [review]
LGTM, just a few nitpicks from me.

File: Gems/MotionMatching/Code/Source/FeatureMatrixStandardScaler.h [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+#pragma once
+
+#include <FeatureMatrix.h>
+#include <FeatureMatrixTransformer.h>
+
+namespace EMotionFX::MotionMatching
+{
+    //! The standard scaler can be used to normalize the feature matrix, the query vector and other data.
+    //! It standardizes features by removing the mean and scaling to the unit variance.
+    //! This implementation is mimicing the behavior of the standard scaler from scikit-learn (sklearn.preprocessing.StandardScaler).
```
[the remark this answers is not in the document]
Comment (amzn-jillich) [reply Gems/MotionMatching/Code/Source/FeatureMatrixStandardScaler.h:18]: Whups, changed :)

File: Gems/MotionMatching/Code/Source/FeatureMatrixStandardScaler.cpp [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+            standardDeviation = 1.0f;
+        }
+
+        // Subtract the mean and scale to unit variance
+        return (value - mean) / standardDeviation;
+    }
+
+    AZ::Vector2 StandardScaler::Transform(const AZ::Vector2& value, FeatureMatrix::Index column) const
+    {
+        return AZ::Vector2(Transform(value.GetX(), column + 0), Transform(value.GetY(), column + 1));
```
Comment (hultonha) [inline Gems/MotionMatching/Code/Source/FeatureMatrixStandardScaler.cpp:73]: question: Here and below is it worth keeping `+ 0`?

File: Gems/MotionMatching/Code/Source/FeatureMatrixStandardScaler.h [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    //! As we use floats by default, our errors are a bigger, especially if the variance is small as this leads to a division by a small
```
[the remark this answers is not in the document]
Comment (amzn-jillich) [reply Gems/MotionMatching/Code/Source/FeatureMatrixStandardScaler.h:19]: Removed, i think that came from originally "a bit" bigger.

File: Gems/MotionMatching/Code/Source/FeatureMatrixStandardScaler.cpp [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+                result(row, column) = InverseTransform(featureMatrix(row, column), column);
+            }
+        }
+
+        return result;
+    }
+
+    AZ::Vector2 StandardScaler::InverseTransform(const AZ::Vector2& value, FeatureMatrix::Index column) const
+    {
+        return AZ::Vector2(InverseTransform(value.GetX(), column + 0), InverseTransform(value.GetY(), column + 1));
```
Comment (hultonha) [inline Gems/MotionMatching/Code/Source/FeatureMatrixStandardScaler.cpp:131]: question: remove `+0` again? 🤔

File: Gems/MotionMatching/Code/Source/FeatureMatrixStandardScaler.h (hunk 1 above)
Comment (amzn-jillich) [reply Gems/MotionMatching/Code/Source/FeatureMatrixStandardScaler.h:62]: Removed as it is currently unused.

File: Gems/MotionMatching/Code/Source/FeatureMatrixStandardScaler.h (hunk 2 above)
Comment (hultonha) [inline Gems/MotionMatching/Code/Source/FeatureMatrixStandardScaler.h:18]: nit: minimicing -> minimizing

File: Gems/MotionMatching/Code/Source/FeatureMatrixStandardScaler.h (hunk 4 above)
Comment (hultonha) [inline Gems/MotionMatching/Code/Source/FeatureMatrixStandardScaler.h:19]: nit: errors are a -> errors are

File: Gems/MotionMatching/Code/Source/FeatureMatrixStandardScaler.h [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+ */
+
+#pragma once
+
+#include <FeatureMatrix.h>
+#include <FeatureMatrixTransformer.h>
+
+namespace EMotionFX::MotionMatching
+{
+    //! The standard scaler can be used to normalize the feature matrix, the query vector and other data.
```
Comment (hultonha) [inline Gems/MotionMatching/Code/Source/FeatureMatrixStandardScaler.h:16]: Great overall comment and description

File: Gems/MotionMatching/Code/Source/FeatureMatrixStandardScaler.h [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    //! roundtrip.
+    class StandardScaler : public FeatureMatrixTransformer
+    {
+    public:
+        AZ_RTTI(StandardScaler, "{A0C7F056-94B0-43A1-8D12-B1A7B67AB92A}", FeatureMatrixTransformer);
+        AZ_CLASS_ALLOCATOR_DECL;
+
+        StandardScaler() = default;
+
+        bool Fit(const FeatureMatrix& featureMatrix, const Settings& settings = {}) override;
```
Comment (hultonha) [inline Gems/MotionMatching/Code/Source/FeatureMatrixStandardScaler.h:31]: suggestion: Api comment for this might be helpful?

File: Gems/MotionMatching/Code/Source/FeatureMatrixStandardScaler.h (hunk 1 above)
Comment (hultonha) [inline Gems/MotionMatching/Code/Source/FeatureMatrixStandardScaler.h:62]: nit: Missing `//!<` (just //)

File: Gems/MotionMatching/Code/Source/MotionMatchingData.h [hunk 8]
```diff
@@ -41,6 +41,12 @@ namespace EMotionFX::MotionMatching
         MotionMatchingData(const FeatureSchema& featureSchema);
         virtual ~MotionMatchingData();
 
+        enum EMFX_API FeatureScalerType
```
Comment (hultonha) [inline Gems/MotionMatching/Code/Source/MotionMatchingData.h:44]: question: is `EMFX_API` needed here for the enum?

**Review by hultonha — APPROVED** [review]
Nice updates! And the tests look great too, my comments were all pretty minor 👍

File: Gems/MotionMatching/Code/Source/FeatureMatrixStandardScaler.cpp [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+        }
+
+        return true;
+    }
+
+    //-------------------------------------------------------------------------
+
+    float StandardScaler::Transform(float value, FeatureMatrix::Index column) const
+    {
+        const float& mean = m_means[column];
```
[the remark this answers is not in the document]
Comment (amzn-jillich) [reply Gems/MotionMatching/Code/Source/FeatureMatrixStandardScaler.cpp:59]: Good call, changed. There was an in-place operation ongoing previously and I forgot to switch back.

File: Gems/MotionMatching/Code/Source/FeatureMatrixStandardScaler.h (hunk 4 above)
Comment (amzn-jillich) [reply Gems/MotionMatching/Code/Source/FeatureMatrixStandardScaler.h:19]: Changed

File: Gems/MotionMatching/Code/Source/FeatureMatrixStandardScaler.h (hunk 7 above)
Comment (amzn-jillich) [reply Gems/MotionMatching/Code/Source/FeatureMatrixStandardScaler.h:31]: Is in the base class already

File: Gems/MotionMatching/Code/Source/FeatureMatrixStandardScaler.h (hunk 1 above)
Comment (amzn-jillich) [reply Gems/MotionMatching/Code/Source/FeatureMatrixStandardScaler.h:62]: Fixed

File: Gems/MotionMatching/Code/Source/MotionMatchingData.h (hunk 8 above)
Comment (amzn-jillich) [reply Gems/MotionMatching/Code/Source/MotionMatchingData.h:44]: Removed

File: Gems/MotionMatching/Code/Source/FeatureMatrixStandardScaler.h (hunk 7 above)
Comment (hultonha) [reply Gems/MotionMatching/Code/Source/FeatureMatrixStandardScaler.h:31]: Aah got you thanks 👍
